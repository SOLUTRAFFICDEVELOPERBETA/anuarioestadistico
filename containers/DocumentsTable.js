import React from 'react'
import PropTypes from 'prop-types'
import {
  makeStyles,
  Paper,
  TableContainer,
  Table,
  TableHead,
  TableCell,
  TableBody,
  TableRow,
  TableSortLabel,
  Toolbar,
  TablePagination,
  Tooltip,
  IconButton
} from '@material-ui/core'
import moment from 'moment'
import { useRouter } from 'next/router'
import SearchInput from '../components/SearchInput'
import { red } from '@material-ui/core/colors'
import { Close, NavigateNext } from '@material-ui/icons'

function descendingComparator (a, b, orderBy) {
  if (b[orderBy] < a[orderBy]) {
    return -1
  }
  if (b[orderBy] > a[orderBy]) {
    return 1
  }
  return 0
}

function getComparator (order, orderBy) {
  return order === 'desc'
    ? (a, b) => descendingComparator(a, b, orderBy)
    : (a, b) => -descendingComparator(a, b, orderBy)
}

function stableSort (array, comparator) {
  const stabilizedThis = array.map((el, index) => [el, index])
  stabilizedThis.sort((a, b) => {
    const order = comparator(a[0], b[0])
    if (order !== 0) return order
    return a[1] - b[1]
  })
  return stabilizedThis.map((el) => el[0])
}

const headCells = [
  { id: 'title', numeric: false, disablePadding: false, label: 'Titulo' },
  { id: 'lastModified', numeric: true, disablePadding: false, label: 'Ultima modificaciÃ³n' }
]

function DocumentTableHead ({ classes, order, orderBy, onRequestSort }) {
  const createSortHandler = (property) => (event) => {
    onRequestSort(event, property)
  }

  return (
    <TableHead>
      <TableRow>
        {headCells.map((headCell) => (
          <TableCell
            key={headCell.id}
            align={headCell.numeric ? 'right' : 'left'}
            padding={headCell.disablePadding ? 'none' : 'default'}
            sortDirection={orderBy === headCell.id ? order : false}
          >
            <TableSortLabel
              active={orderBy === headCell.id}
              direction={orderBy === headCell.id ? order : 'asc'}
              onClick={createSortHandler(headCell.id)}
            >
              {headCell.label}
              {orderBy === headCell.id ? (
                <span className={classes.visuallyHidden}>
                  {order}
                </span>
              ) : null}
            </TableSortLabel>
          </TableCell>
        ))}
        <TableCell />
      </TableRow>
    </TableHead>
  )
}

DocumentTableHead.propTypes = {
  classes: PropTypes.object.isRequired,
  onRequestSort: PropTypes.func.isRequired,
  order: PropTypes.oneOf(['asc', 'desc']).isRequired,
  orderBy: PropTypes.string.isRequired
}

const useToolbarStyles = makeStyles((theme) => ({
  root: {
    paddingLeft: theme.spacing(2),
    paddingRight: theme.spacing(1)
  },
  title: {
    flex: '1 1 100%'
  }
}))

const DocumentTableToolbar = ({ onSearch }) => {
  const classes = useToolbarStyles()

  return (
    <Toolbar
      className={classes.root}
    >
      <div className={classes.title}>
        <SearchInput placeholder="Buscar por nombre" onSearch={onSearch} />
      </div>
    </Toolbar>
  )
}

DocumentTableToolbar.propTypes = {
  onSearch: PropTypes.func.isRequired
}

const useStyles = makeStyles(theme => ({
  root: {
    width: '100%'
  },
  paper: {
    width: '100%',
    margin: theme.spacing(1, 0)
  },
  table: {
    minWidth: 750
  },
  visuallyHidden: {
    border: 0,
    clip: 'rect(0 0 0 0)',
    height: 1,
    margin: -1,
    overflow: 'hidden',
    padding: 0,
    position: 'absolute',
    top: 20,
    width: 1
  },
  delete: {
    backgroundColor: red[500],
    color: theme.palette.getContrastText(red[500]),
    marginRight: theme.spacing(1),

    '&:hover': {
      color: red[500]
    }
  },
  send: {
    backgroundColor: theme.palette.primary.main,
    color: theme.palette.primary.contrastText,

    '&:hover': {
      color: theme.palette.primary.main
    }
  }
}))

const DocumentsTable = ({ documents, onDelete }) => {
  const classes = useStyles()
  const router = useRouter()
  const [docs, setDocs] = React.useState([...documents])
  const [order, setOrder] = React.useState('desc')
  const [orderBy, setOrderBy] = React.useState('lastModified')
  const [page, setPage] = React.useState(0)
  const [rowsPerPage, setRowsPerPage] = React.useState(5)
  const { query: { id } } = router

  const handleRequestSort = (event, property) => {
    const isAsc = orderBy === property && order === 'asc'
    setOrder(isAsc ? 'desc' : 'asc')
    setOrderBy(property)
  }

  const handleChangePage = (event, newPage) => {
    setPage(newPage)
  }

  const handleChangeRowsPerPage = (event) => {
    setRowsPerPage(parseInt(event.target.value, 10))
    setPage(0)
  }

  const handleSearch = ({ search }) => {
    const array = [...documents].filter(c => c.title.toLowerCase().includes(search.toLowerCase()))
    setDocs(array)
  }

  const emptyRows = rowsPerPage - Math.min(rowsPerPage, docs.length - page * rowsPerPage)

  React.useEffect(() => {
    setDocs(documents)
  }, [documents])

  return (
    <div className={classes.root}>
      <Paper className={classes.paper}>
        <DocumentTableToolbar onSearch={handleSearch} />
        <TableContainer>
          <Table className={classes.table}>
            <DocumentTableHead
              classes={classes}
              order={order}
              orderBy={orderBy}
              onRequestSort={handleRequestSort}
            />
            <TableBody>
              {stableSort(docs, getComparator(order, orderBy))
                .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                .map((row, index) => {
                  return (
                    <TableRow
                      hover
                      role="checkbox"
                      tabIndex={-1}
                      key={row.name}
                    >
                      <TableCell scope="row">{row.title}</TableCell>
                      <TableCell scope="row" align="right">{moment(row.lastModified).format('ll')}</TableCell>
                      <TableCell scope="row" align="center" padding="none">
                        <Tooltip title="Borrar"  arrow>
                          <IconButton size="small" className={classes.delete} onClick={() => onDelete(row.id)}>
                            <Close fontSize="small"/>
                          </IconButton>
                        </Tooltip>
                        <Tooltip title="Abrir" arrow>
                          <IconButton
                            size="small"
                            className={classes.send}
                            onClick={() => router.push('/paginas/createPages/[id]', `/paginas/createPages/${row.id}`)}
                          >
                            <NavigateNext fontSize="small"  />
                          </IconButton>
                        </Tooltip>
                      </TableCell>
                    </TableRow>
                  )
                })}
              {emptyRows > 0 && (
                <TableRow style={{ height: 53 * emptyRows }}>
                  <TableCell colSpan={6} />
                </TableRow>
              )}
            </TableBody>
          </Table>
        </TableContainer>
        <TablePagination
          rowsPerPageOptions={[5, 10, 25]}
          component="div"
          count={docs.length}
          rowsPerPage={rowsPerPage}
          page={page}
          onChangePage={handleChangePage}
          onChangeRowsPerPage={handleChangeRowsPerPage}
        />
      </Paper>
    </div>
  )
}

DocumentsTable.propTypes = {
  documents: PropTypes.array,
  onDelete: PropTypes.func
}

export default DocumentsTable
