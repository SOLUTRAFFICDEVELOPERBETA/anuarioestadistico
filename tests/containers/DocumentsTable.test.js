import React from 'react'
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import moment from 'moment';
import DocumentsTable from '../../containers/DocumentsTable';
import { documents } from '../fixtures';


describe('Pruebas en <DocumentsTable/>', () => {
    const onDelete = jest.fn();
    const wrapper = shallow(
        <DocumentsTable
            documents={documents}
            onDelete={onDelete}
        />
    )
    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    })
    test('dede mostrar la información enviada correctamente', () => {
        const title = wrapper.find('WithStyles(ForwardRef(TableCell))').at(0);
        const date = wrapper.find('WithStyles(ForwardRef(TableCell))').at(1);
        expect(title.text().trim()).toBe(documents[0].title);
        expect(date.text().trim()).toBe(moment(documents[0].lastModified).format('ll'));
    })

    test('debe llamar la función onDelete() ', () => {
        wrapper.find('WithStyles(ForwardRef(IconButton))').at(0).prop('onClick')();
        expect(onDelete).toHaveBeenCalled();
        expect(onDelete).toHaveBeenCalledTimes(1);
    })
})