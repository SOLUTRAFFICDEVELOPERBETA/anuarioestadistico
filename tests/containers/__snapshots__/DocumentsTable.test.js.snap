// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <DocumentsTable/> debe mostrarse correctamente 1`] = `
<div
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-2"
  >
    <DocumentTableToolbar
      onSearch={[Function]}
    />
    <WithStyles(ForwardRef(TableContainer))>
      <WithStyles(ForwardRef(Table))
        className="makeStyles-table-3"
      >
        <DocumentTableHead
          classes={
            Object {
              "delete": "makeStyles-delete-5",
              "paper": "makeStyles-paper-2",
              "root": "makeStyles-root-1",
              "send": "makeStyles-send-6",
              "table": "makeStyles-table-3",
              "visuallyHidden": "makeStyles-visuallyHidden-4",
            }
          }
          onRequestSort={[Function]}
          order="desc"
          orderBy="lastModified"
        />
        <WithStyles(ForwardRef(TableBody))>
          <WithStyles(ForwardRef(TableRow))
            hover={true}
            key="name for default"
            role="checkbox"
            tabIndex={-1}
          >
            <WithStyles(ForwardRef(TableCell))
              scope="row"
            >
              Titulo de pagina de prueba UNO
            </WithStyles(ForwardRef(TableCell))>
            <WithStyles(ForwardRef(TableCell))
              align="right"
              scope="row"
            >
              Dec 16, 2020
            </WithStyles(ForwardRef(TableCell))>
            <WithStyles(ForwardRef(TableCell))
              align="center"
              padding="none"
              scope="row"
            >
              <WithStyles(ForwardRef(Tooltip))
                arrow={true}
                title="Borrar"
              >
                <WithStyles(ForwardRef(IconButton))
                  className="makeStyles-delete-5"
                  onClick={[Function]}
                  size="small"
                >
                  <CloseIcon
                    fontSize="small"
                  />
                </WithStyles(ForwardRef(IconButton))>
              </WithStyles(ForwardRef(Tooltip))>
              <WithStyles(ForwardRef(Tooltip))
                arrow={true}
                title="Abrir"
              >
                <WithStyles(ForwardRef(IconButton))
                  className="makeStyles-send-6"
                  onClick={[Function]}
                  size="small"
                >
                  <NavigateNextIcon
                    fontSize="small"
                  />
                </WithStyles(ForwardRef(IconButton))>
              </WithStyles(ForwardRef(Tooltip))>
            </WithStyles(ForwardRef(TableCell))>
          </WithStyles(ForwardRef(TableRow))>
          <WithStyles(ForwardRef(TableRow))
            hover={true}
            key="name for default"
            role="checkbox"
            tabIndex={-1}
          >
            <WithStyles(ForwardRef(TableCell))
              scope="row"
            >
              Titulo de pagina de prueba DOS
            </WithStyles(ForwardRef(TableCell))>
            <WithStyles(ForwardRef(TableCell))
              align="right"
              scope="row"
            >
              Dec 16, 2020
            </WithStyles(ForwardRef(TableCell))>
            <WithStyles(ForwardRef(TableCell))
              align="center"
              padding="none"
              scope="row"
            >
              <WithStyles(ForwardRef(Tooltip))
                arrow={true}
                title="Borrar"
              >
                <WithStyles(ForwardRef(IconButton))
                  className="makeStyles-delete-5"
                  onClick={[Function]}
                  size="small"
                >
                  <CloseIcon
                    fontSize="small"
                  />
                </WithStyles(ForwardRef(IconButton))>
              </WithStyles(ForwardRef(Tooltip))>
              <WithStyles(ForwardRef(Tooltip))
                arrow={true}
                title="Abrir"
              >
                <WithStyles(ForwardRef(IconButton))
                  className="makeStyles-send-6"
                  onClick={[Function]}
                  size="small"
                >
                  <NavigateNextIcon
                    fontSize="small"
                  />
                </WithStyles(ForwardRef(IconButton))>
              </WithStyles(ForwardRef(Tooltip))>
            </WithStyles(ForwardRef(TableCell))>
          </WithStyles(ForwardRef(TableRow))>
          <WithStyles(ForwardRef(TableRow))
            style={
              Object {
                "height": 159,
              }
            }
          >
            <WithStyles(ForwardRef(TableCell))
              colSpan={6}
            />
          </WithStyles(ForwardRef(TableRow))>
        </WithStyles(ForwardRef(TableBody))>
      </WithStyles(ForwardRef(Table))>
    </WithStyles(ForwardRef(TableContainer))>
    <WithStyles(ForwardRef(TablePagination))
      component="div"
      count={2}
      onChangePage={[Function]}
      onChangeRowsPerPage={[Function]}
      page={0}
      rowsPerPage={5}
      rowsPerPageOptions={
        Array [
          5,
          10,
          25,
        ]
      }
    />
  </WithStyles(ForwardRef(Paper))>
</div>
`;
